File: js/src/Node.js
// When this library loaded, this will extends Blackprint's internal to use this library for syncing nodes
// Used for sending node's data from Sketch -> Engine, or Engine -> Sketch

Blackprint.Node.prototype.syncOut = function(id, data=''){
	let instance = this.instance.rootInstance || this.instance;
	if(instance._remote == null || this._syncronizing || instance.syncDataOut === false)
		return;

	if(id.constructor !== String)
		throw new Error("syncOut's ID must be a string, but got: "+ id);

	let char = id.slice(0, 1);
	if(char === '_' || char === '$'){
		throw new Error("syncOut's ID can't be started with '_' or '$' character as it's assumed as a private field, but got: "+ id);
	}

	if(this.syncThrottle !== 0){
		clearTimeout(this._syncHasWait);
		this._syncHasWait = setTimeout(() => {
			if(this._syncHasWait)
				instance.emit('_node.sync', {
					iface: this.iface,
					data: clearPrivateField(this._syncWait)
				});

			this._syncWait = null;
		}, this.syncThrottle);

		if(this._syncWait == null)
			this._syncWait = {};

		this._syncWait[id] = data;
	}
	else instance.emit('_node.sync', {iface: this.iface, data: clearPrivateField({ [id]: data })});
}

function clearPrivateField(obj){
	if(obj == null) return obj;

	if(obj instanceof Array){
		let temp = obj.slice(0);
		for (var i = 0; i < temp.length; i++) {
			let ref = temp[i];

			if(typeof ref === 'object')
				temp[i] = clearPrivateField(ref);
		}

		return temp;
	}

	let temp = {};
	for(let key in obj){
		let char = key.slice(0, 1);
		if(char === '_' || char === '$')
			continue;

		let ref = obj[key];

		if(typeof ref === 'object')
			temp[key] = clearPrivateField(ref);
		else temp[key] = ref;
	}

	return temp;
}

File: lua/src/Node.lua
local Node = {}
local Timer = require("@src/LuaUtils/Timer.lua")

function Node.BpSyncOut(node, id, data, force)
	local instance = node.instance.rootInstance or node.instance

	if instance._remote == nil or (not force and node._syncronizing) or instance.syncDataOut == false then
		return
	end

	if type(id) ~= "string" then
		error("syncOut's ID must be a string, but got: " .. type(id))
	end

	local char = id:sub(1, 1)
	if char == '_' or char == '$' then
		error("syncOut's ID can't be started with '_' or '$' character as it's assumed as a private field, but got: " .. id)
	end

	if node.syncThrottle ~= 0 then -- ToDo: make this timer more efficient
		if node._syncWait == nil then node._syncWait = {} end
		node._syncWait[id] = data

		local function timeout()
			if node._syncHasWait then
				instance:emit('_node.sync', {
					iface = node.iface,
					data = Node.clearPrivateField(node._syncWait)
				})
			end

			node._syncWait = nil
			node._syncHasWait = false
		end

		if node._syncHasWait then
			Timer.clearTimeout(node._syncHasWait)
		end
		node._syncHasWait = Timer.setTimeout(timeout, node.syncThrottle)
	else
		instance:emit('_node.sync', {
			iface = node.iface,
			data = Node.clearPrivateField({ [id] = data })
		})
	end
end

function Node.clearPrivateField(obj)
	if obj == nil then return obj end

	if type(obj) == "table" then
		local temp = {}
		for key, value in pairs(obj) do
			local char = key:sub(1, 1)
			if char == '_' or char == '$' then
				continue
			end

			if type(value) == "table" then
				temp[key] = Node.clearPrivateField(value)
			else
				temp[key] = value
			end
		end
		return temp
	end

	return obj
end

return Node